blockchain/process.go

```diff
@@ -47,48 +365,219 @@ const (
 
-	// The block has passed all context independent checks and appears sane
-	// enough to potentially accept it into the block chain.
-	forkLen, err := b.maybeAcceptBlock(block, flags)
+	// Enable skipping some of the more expensive validation checks when the
+	// block is an ancestor of a known good checkpoint.
+	//
+	// NOTE: The fast add flag from the caller is used as a standin for now
+	// since that validation is currently handled by the calling code.  In the
+	// future, the calling code should be updated to process all headers through
+	// this pakcage and then the ability to specify the fast add flag should be
+	// removed along with the fast add portion of this check here so that it is
+	// solely determined internally.
+	if flags&BFFastAdd == BFFastAdd || b.isKnownCheckpointAncestor(node) {
+		b.index.SetStatusFlags(node, statusValidated)
+		flags |= BFFastAdd
+	}
+
+	// Potentially accept the block data into the database and update the block
+	// index state to account for the full data now being available.
+	//
+	// This consists of performing several validation checks which depend on the
+	// block's position within the block chain and determining if the block, and
+	// any descendants of it are now eligible for full validation due to being
+	// fully linked (meaning a block has all of its own data available and all
+	// of its ancestors also have their data available).
+	//
+	// The returned linked block nodes are for those aforementioned blocks that
+	// are now eligible for validation.
+	linkedNodes, err := b.maybeAcceptBlockData(node, block, flags)
 	if err != nil {
 		return 0, err
 	}
 
-	log.Debugf("Accepted block %v", blockHash)
+	// Write any modified block index entries to the database since any new
+	// headers will have added a new entry and the block will be marked as now
+	// having its data stored.
+	if err := b.flushBlockIndex(); err != nil {
+		return 0, err
+	}
+
+	// Tentatively accept the linked blocks to the chain if they pass several
+	// validation checks which depend on having the full block data for all of
+	// their ancestors available and update the block index state to account for
+	// any that fail validation.
+	//
+	// Note that this done here because it allows any blocks that fail this
+	// level of validation to be detected and discounted early before doing more
+	// work.
+	//
+	// Also, any blocks that do not ultimately end up becoming part of the best
+	// chain would otherwise not have contextual checks run on them, which is
+	// required before accepting them, without somewhat more complicated logic
+	// later to detect them.
+	var finalErr error
+	currentTip := b.bestChain.Tip()
+	b.addRecentBlock(block)
+	acceptedNodes, err := b.maybeAcceptBlocks(currentTip, linkedNodes, flags)
+	if err != nil {
+		finalErr = err
 
-	return forkLen, nil
+		// This intentionally falls through since the code below must run to
+		// account for any blocks that were accepted even when not all of them
+		// were.
```

Comment:
No, I don't believe so because it is possible the daemon was interrupted while connecting blocks previously and thus the current tip might actually be behind the latest possible block which the reorganize below will handle to catch up too.

Reply:
Makes sense, thanks for the clarification.  That comment may be slightly misleading then since the code below should run even if _none_ were accepted (by reading only the code I would have had the same conclusion as you but the comment led me to think otherwise).

Reply:
Good point.  I've updated the comment to better call that out.

 		// This intentionally falls through since the code below must run
		// whether or not any blocks were accepted.